src/umpire/resource/HostMpi3SharedMemoryResource.cpp

```diff
@@ -0,0 +1,66 @@
+//////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2016-24, Lawrence Livermore National Security, LLC and Umpire
+// project contributors. See the COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (MIT)
+//////////////////////////////////////////////////////////////////////////////
+
+#include "umpire/resource/HostMpi3SharedMemoryResource.hpp"
+
+#include "umpire/resource/MemoryResource.hpp"
+#include "umpire/util/Macros.hpp"
+#include "umpire/util/error.hpp"
+
+namespace umpire {
+namespace resource {
+
+HostMpi3SharedMemoryResource::HostMpi3SharedMemoryResource(const std::string& name, int id, MemoryResourceTraits traits)
+    : MemoryResource{name, id, traits}
+{
+  constexpr int IGNORE_KEY{0};
+  MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_SHARED, IGNORE_KEY, MPI_INFO_NULL, &m_shared_comm);
+  MPI_Comm_rank(m_shared_comm, &m_local_rank);
+}
+
+HostMpi3SharedMemoryResource::~HostMpi3SharedMemoryResource()
+{
+  MPI_Comm_free(&m_shared_comm);
+}
+
+void* HostMpi3SharedMemoryResource::allocate(std::size_t bytes)
+{
+  void* ptr;
+  MPI_Win win;
+  MPI_Aint size = (m_local_rank != 0) ? 0 : bytes;
+  int disp{sizeof(char)};
+
+  MPI_Win_allocate_shared(size, disp, MPI_INFO_NULL, m_shared_comm, &ptr, &win);
+
```

Comment:
Isn't MPI_Win_shared_query needed if m_local_rank != 0?

Reply:
That's for "non-contiguous" shared memory.

Reply:
So the following statement that website made is incorrect? Pardon my lack of knowledge about MPI shared memory.

"You need this method for non-contiguous shared memory, but also for all processes
with a zero byte window portion, because if a process specifies in MPI_Win_allocate_shared a zero byte window length, then no base pointer is returned."

Reply:
Or this statement from the manpage for MPI_Win_allocate_shared?

"Contiguous across process ranks means that the first address in the memory segment of process i is consecutive with the last address in the memory segment of process i - 1. This may enable the user to calculate remote address offsets with local information only."

Reply:
I think you are correct. The documentation is difficult to follow.

Reply:
I wasn't expecting so many review comments yet - PR still needs tests written :)

Reply:
Haha I've been wanting this feature for a while :)